_posts/2022-10-27-pytorch-experiment-tracking.md

```diff
@@ -0,0 +1,657 @@
+    print(f"Epoch {t+1}\n-------------------------------")
+    train(train_dataloader, model, loss_fn, optimizer)
+    test(test_dataloader, model, loss_fn)
+print("Done!")
+
+```
+
+You can put all the codes into a `.py` file, and it should run without a problem.
+
+Now let's see how we can transform the codes into a ZenML pipeline.
+
+
+## 🥳 Transforming PyTorch Codes into a ZenML Pipeline.
+Before we start, I'd like to first tell you about the concept of *pipeline* and *step* in ZenML. This concept will come in handy later when we code.
+
+In ZenML, a `pipeline` consists of a series of steps, organized in any order that makes sense for your use case.
+
+The following illustration is a simple `pipeline` that consists of three `steps` running one after another.
+
+![pipeline_steps](/assets/posts/pytorch_wandb/pipeline_step.gif)
+
+The figure above is the exact `pipeline` and `steps` that we will construct from the vanilla PyTorch code.
+Let's start the transformation.
+
+First, import modules from `torch`, `torchvision` and `zenml`.
+
+```python
+import torch
+from torch import nn
+from torch.utils.data import DataLoader
+from torchvision import datasets
+from torchvision.transforms import ToTensor
+
+from zenml.pipelines import pipeline
+from zenml.steps import step, Output
+```
+
+Next, let's define the `pipeline`.
+
+You can do this by putting a `@pipeline` decorator above the function definition.
+
+```python
+@pipeline
+def pytorch_experiment_tracking_pipeline(
+    load_data,
+    load_model,
+    train_test,
+):
+    """A `pipeline` to load data, load model, and train/evaluate the model."""
+    train_dataloader, test_dataloader = load_data()
+    model = load_model()
+    train_test(model, train_dataloader, test_dataloader)
+```
+The pipeline we just wrote takes three `steps` as the input namely - `load_data`, `load_model`, and `train_test`. 
+Each `step` runs sequentially one after another. 
+
+Next, let's define what the `steps` actually do. 
+We can define a `step` in the same way we define a `pipeline`, except we put a `@step` decorator now.
+
+Let's start with the first `step` to load the data.
+
+```python
+@step
+def load_data() -> Output(
+    train_dataloader=DataLoader, test_dataloader=DataLoader
+):
+    """A `step` to load the Fashion MNIST dataset as a tuple of torch Datasets."""
+    batch_size = 64
+
+    # Download training data from open datasets.
+    training_data = datasets.FashionMNIST(
+        root="data",
+        train=True,
+        download=True,
+        transform=ToTensor(),
+    )
+
+    # Download test data from open datasets.
+    test_data = datasets.FashionMNIST(
+        root="data",
+        train=False,
+        download=True,
+        transform=ToTensor(),
+    )
+
+    # Create data loaders.
+    train_dataloader = DataLoader(training_data, batch_size=batch_size)
+    test_dataloader = DataLoader(test_data, batch_size=batch_size)
+
+    return train_dataloader, test_dataloader
```

Comment:
Generally speaking, is returning dataloaders (which are lazy objects) the best way to showcase? Im still unsure about this part in ZenML

Reply:
Ok then lets go for it!